`youtube-dl` requires missing `bin/python3` instead of `bin/python3.12`

### `brew gist-logs <formula>` link OR `brew config` AND `brew doctor` output

```shell
$ brew config
HOMEBREW_VERSION: 4.1.21
ORIGIN: https://github.com/Homebrew/brew
HEAD: ad1bb178d4a478840c4524063ac5cd52d8667d25
Last commit: 5 days ago
Core tap JSON: 23 Nov 18:33 UTC
HOMEBREW_PREFIX: /opt/homebrew
HOMEBREW_BAT_THEME: ansi
HOMEBREW_CASK_OPTS: []
HOMEBREW_DISPLAY: /private/tmp/com.apple.launchd.0ZgBKG07mS/org.xquartz:0
HOMEBREW_MAKE_JOBS: 10
HOMEBREW_NO_ANALYTICS: set
Homebrew Ruby: 2.6.10 => /System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/bin/ruby
CPU: 10-core 64-bit arm_firestorm_icestorm
Clang: 14.0.3 build 1403
Git: 2.43.0 => /opt/homebrew/bin/git
Curl: 7.88.1 => /usr/bin/curl
macOS: 13.4-arm64
CLT: 14.3.1.0.1.1683849156
Xcode: 14.3.1
Rosetta 2: false


$ brew doctor
Please note that these warnings are just used to help the Homebrew maintainers
with debugging if you file an issue. If everything you use Homebrew for is
working fine: please don't worry or file an issue; just ignore this. Thanks!

Warning: Some installed formulae are deprecated or disabled.
You should find replacements for the following formulae:
  cling
  openssl@1.1
```


### Verification

- [X] My "`brew doctor` output" says `Your system is ready to brew.` and am still able to reproduce my issue.
- [X] I ran `brew update` and am still able to reproduce my issue.
- [X] I have resolved all warnings from `brew doctor` and that did not fix my problem.
- [X] I searched for recent similar issues at https://github.com/Homebrew/homebrew-core/issues?q=is%3Aissue and found no duplicates.

### What were you trying to do (and why)?

To install an up to date `youtube-dl`.

The failure seems to be that it tries to use `/opt/homebrew/opt/python@3.12/bin/python3` but that file does not exist. It probably should want `/opt/homebrew/opt/python@3.12/bin/python3.12`.

I tried symlinking `/opt/homebrew/opt/python@3.12/bin/python3.12 -> /opt/homebrew/opt/python@3.12/bin/python3` and everything worked perfectly; I imagine this is a config option somewhere. Maybe here: https://github.com/Homebrew/homebrew-core/blob/adb653972409c4cfc269cd9ae0d92e27cc0c670b/Formula/y/youtube-dl.rb#L30

As per brew's instructions I did first make an issue with the package itself but they directed me back here: https://github.com/ytdl-org/youtube-dl/issues/32644#issuecomment-1824974988

If it matters, my `which python3` is `/opt/homebrew/bin/python3` which is a symlink to `python3.11`, not `python3.12`.

### What happened (include all command output)?

```
brew install --HEAD youtube-dl                                                                          1
==> Downloading https://formulae.brew.sh/api/formula.jws.json
################################################################################################################### 100.0%
==> Downloading https://formulae.brew.sh/api/cask.jws.json
################################################################################################################### 100.0%
==> Fetching dependencies for youtube-dl: readline
==> Fetching readline
==> Downloading https://ghcr.io/v2/homebrew/core/readline/manifests/8.2.7
################################################################################################################### 100.0%
==> Downloading https://ghcr.io/v2/homebrew/core/readline/blobs/sha256:1cbe9a00128c3b5cdf1107419087f00505533c73a60c3a21021
################################################################################################################### 100.0%
==> Fetching youtube-dl
==> Downloading https://raw.githubusercontent.com/Homebrew/homebrew-core/6bf10ccb88ed119157d7fda1adcbd0aaa135a925/Formula/
################################################################################################################### 100.0%
==> Cloning https://github.com/ytdl-org/youtube-dl.git
Updating /Users/zwimer/Library/Caches/Homebrew/youtube-dl--git
==> Checking out branch master
Already on 'master'
Your branch is up to date with 'origin/master'.
HEAD is now at aaed4884e fix(youtube): update extractor handling consent
==> Installing dependencies for youtube-dl: readline
==> Installing youtube-dl dependency: readline
==> Downloading https://ghcr.io/v2/homebrew/core/readline/manifests/8.2.7
Already downloaded: /Users/zwimer/Library/Caches/Homebrew/downloads/14125f7fa4b49853f76160b864f58379d90e52833ffeb8bd0643609bcd7f02a7--readline-8.2.7.bottle_manifest.json
==> Pouring readline--8.2.7.arm64_ventura.bottle.tar.gz
🍺  /opt/homebrew/Cellar/readline/8.2.7: 50 files, 1.7MB
==> Installing youtube-dl --HEAD
==> make PREFIX=/opt/homebrew/Cellar/youtube-dl/HEAD-aaed488 MANDIR=/opt/homebrew/Cellar/youtube-dl/HEAD-aaed488/share/man
Last 15 lines from /Users/zwimer/Library/Logs/Homebrew/youtube-dl/01.make:
/bin/sh: /opt/homebrew/opt/python@3.12/bin/python3: No such file or directory
make: *** [youtube-dl.bash-completion] Error 1
make: *** Waiting for unfinished jobs....
/bin/sh: /opt/homebrew/opt/python@3.12/bin/python3: No such file or directory
make: *** [youtube-dl.zsh] Error 1
make: *** [README.md] Error 127
make: *** [youtube-dl.fish] Error 1
touch -t 200001010101 zip/youtube_dl/*.py zip/youtube_dl/*/*.py
mv zip/youtube_dl/__main__.py zip/
cd zip ; zip -q ../youtube-dl youtube_dl/*.py youtube_dl/*/*.py __main__.py
rm -rf zip
echo '#!/opt/homebrew/opt/python@3.12/bin/python3' > youtube-dl
cat youtube-dl.zip >> youtube-dl
rm youtube-dl.zip
chmod a+x youtube-dl

READ THIS: https://docs.brew.sh/Troubleshooting

youtube-dl's formula was built from an unstable upstream --HEAD.
This build failure is expected behaviour.
Do not create issues about this on Homebrew's GitHub repositories.
Any opened issues will be immediately closed without response.
Do not ask for help from Homebrew or its maintainers on social media.
You may ask for help in Homebrew's discussions but are unlikely to receive a response.
Try to figure out the problem yourself and submit a fix as a pull request.
We will review it but may or may not accept it.
```

### What did you expect to happen?

It to install correctly, like it does if I symlink: `/opt/homebrew/opt/python@3.12/bin/python3.12 -> /opt/homebrew/opt/python@3.12/bin/python3`

### Step-by-step reproduction instructions (by running `brew` commands)

```shell
brew install --HEAD youtube-dl
```

--- a/Formula/y/youtube-dl.rb
+++ b/Formula/y/youtube-dl.rb
@@ -17,14 +17,14 @@ class YoutubeDl < Formula
   end
 
   bottle do
-    rebuild 3
-    sha256 cellar: :any_skip_relocation, arm64_sonoma:   "91ddcd5a78a536d49b1dd461b1f841611cbf641fae881aa4b88fd354451d9848"
-    sha256 cellar: :any_skip_relocation, arm64_ventura:  "a5c07f4bf1dba666a29a383db3a93e999e21e87cbd897247fe90bd48bbbd6629"
-    sha256 cellar: :any_skip_relocation, arm64_monterey: "2e8e5b503f45df8295c549bd1b3c8852eddeb1a820ef83609346ff05eb0aef42"
-    sha256 cellar: :any_skip_relocation, sonoma:         "aa5eed9e9d1c14f924f90b67c653cc0207e22bb86ee3280fb5e9f114383c7bd0"
-    sha256 cellar: :any_skip_relocation, ventura:        "b05a5de77bed702e043fa611811597321b92c15b3d8ee5d7abc07db452966219"
-    sha256 cellar: :any_skip_relocation, monterey:       "5c53c7c184fe002c221ce12eef1dde7aba6dc200b8f63a2fb55db1c2059ae85b"
-    sha256 cellar: :any_skip_relocation, x86_64_linux:   "db272dcc243e7d6ad084e1c63e407379be8501eb7727f13445b04ea0bd5dffc8"
+    rebuild 4
+    sha256 cellar: :any_skip_relocation, arm64_sonoma:   "86ad6daf8ba59b37855706201030abfdfa6a334d66ee80ee6351212baa1fccaa"
+    sha256 cellar: :any_skip_relocation, arm64_ventura:  "0f0777822da37695c41651a1745fda40a3bda17b7d252b682c74eaa8591023b5"
+    sha256 cellar: :any_skip_relocation, arm64_monterey: "9e1ac5fa2ec5d0dc609a2a52b5862da268bcd2731609aa114491b787105701f1"
+    sha256 cellar: :any_skip_relocation, sonoma:         "d203c17f40f2e3ab8e1d3887539ae66173c4e91b36d28024e7bd98e41c2407a3"
+    sha256 cellar: :any_skip_relocation, ventura:        "41ea7b884710606f9e4b2e98d186815d1be48a13fa6e3c945c2c686fec48e25f"
+    sha256 cellar: :any_skip_relocation, monterey:       "f78df8ae4f3c7b6f84f8d2a94bfb1187219c0b2c46a8a2859bf1061940068820"
+    sha256 cellar: :any_skip_relocation, x86_64_linux:   "2abdf78bc998a6fde5c66488e36411df9c0d8326554a031c3dd32aaef651af2c"
   end
 
   head do

--- a/Formula/y/youtube-dl.rb
+++ b/Formula/y/youtube-dl.rb
@@ -3,9 +3,18 @@ class YoutubeDl < Formula
 
   desc "Download YouTube videos from the command-line"
   homepage "https://ytdl-org.github.io/youtube-dl/"
-  url "https://files.pythonhosted.org/packages/01/4f/ab0d0806f4d818168d0ec833df14078c9d1ddddb5c42fa7bfb6f15ecbfa7/youtube_dl-2021.12.17.tar.gz"
-  sha256 "bc59e86c5d15d887ac590454511f08ce2c47698d5a82c27bfe27b5d814bbaed2"
   license "Unlicense"
+
+  stable do
+    url "https://files.pythonhosted.org/packages/01/4f/ab0d0806f4d818168d0ec833df14078c9d1ddddb5c42fa7bfb6f15ecbfa7/youtube_dl-2021.12.17.tar.gz"
+    sha256 "bc59e86c5d15d887ac590454511f08ce2c47698d5a82c27bfe27b5d814bbaed2"
+
+    # Backport fix for extractor handling consent
+    patch do
+      url "https://github.com/ytdl-org/youtube-dl/commit/aaed4884ed9954b8b69c3ca5254418ec578ed0b9.patch?full_index=1"
+      sha256 "3078402768839f4ad611bcb7ab3b221d1a97626c62c4f1bdb2f85598fa45fa96"
+    end
+  end
 
   bottle do
     rebuild 3
@@ -26,16 +35,16 @@ class YoutubeDl < Formula
   depends_on "python@3.12"
 
   def install
+    python3 = which("python3.12")
     if build.head?
-      python = Formula["python@3.12"].opt_bin/"python3"
-      system "make", "PREFIX=#{prefix}", "MANDIR=#{man}", "PYTHON=#{python}", "install"
+      system "make", "PREFIX=#{prefix}", "MANDIR=#{man}", "PYTHON=#{python3}", "install"
       fish_completion.install prefix/"etc/fish/completions/youtube-dl.fish"
       (prefix/"etc/fish").rmtree
     else
       virtualenv_install_with_resources
       # Handle "ERROR: Unable to extract uploader id" until new release
       # https://github.com/ytdl-org/youtube-dl/issues/31530
-      inreplace libexec/"lib/python3.12/site-packages/youtube_dl/extractor/youtube.py",
+      inreplace libexec/Language::Python.site_packages(python3)/"youtube_dl/extractor/youtube.py",
                 "owner_profile_url, 'uploader id')",
                 "owner_profile_url, 'uploader id', fatal=False)"
       man1.install_symlink libexec/"share/man/man1/youtube-dl.1" => "youtube-dl.1"

